web3-functions/index.ts

```diff
@@ -0,0 +1,293 @@
+
+const ORACLE_ABI = [
+  "function updateDataFeedsValuesPartial(bytes32[]) public",
+  "function getLastUpdateDetails(bytes32) public view returns (uint256, uint256, uint256)",
+  "function getLivePricesAndTimestamp(bytes32[]) public view returns (uint256[], uint256)",
+];
+
+const MIN_DEVIATION = 0.5; // 0.5%
+const MIN_TIME_ELAPSED = 6; // 6 hours
+const DECIMALS = 8; // price feed precision
+const zero = BigNumber.from(0);
+const one = BigNumber.from(1);
+
+const DEBUG_MODE = false;
+const debugLog = conditionalLog(DEBUG_MODE);
+
+interface DataFeed {
+  symbol: string;
+  id: string;
+  livePrice: BigNumber;
+  storedPrice: BigNumber;
+  storedTimestamp: number;
+}
+
+Web3Function.onRun(async (context: Web3FunctionContext) => {
+  const { userArgs, multiChainProvider } = context;
+
+  const provider = multiChainProvider.default();
+
+  const dataServiceId = userArgs.dataServiceId as string;
+  const dataFeedIdsString = userArgs.symbols as string[];
+  const dataFeedIdsBytes32 = dataFeedIdsString.map((id) =>
+    formatBytes32String(id)
+  );
+  const oracleAddress = userArgs.oracleAddress as string;
+
+  const oracle = new Contract(oracleAddress, ORACLE_ABI, provider);
+
+  let dataFeedIds = new Map<string, DataFeed>();
+  for (const id of dataFeedIdsString) {
+    dataFeedIds.set(id, {
+      symbol: id,
+      id: formatBytes32String(id),
+      livePrice: BigNumber.from(0),
+      storedPrice: BigNumber.from(0),
+      storedTimestamp: 0,
+    });
+  }
+  debugLog("Data feed ids: ", dataFeedIds);
+
+  // Wrap contract with redstone data service
+  let wrappedOracle;
+  switch (dataServiceId) {
+    case "redstone-primary-prod":
+      wrappedOracle = WrapperBuilder.wrap(oracle).usingDataService(
+        {
+          dataServiceId: "redstone-primary-prod",
+          uniqueSignersCount: 2,
+          dataFeeds: dataFeedIdsString,
+          disablePayloadsDryRun: true,
+        },
+        [
+          "https://oracle-gateway-1.a.redstone.finance",
+          "https://oracle-gateway-2.a.redstone.finance",
+        ]
+      );
+      break;
+    case "redstone-main-demo":
+      wrappedOracle = WrapperBuilder.wrap(oracle).usingDataService(
+        {
+          dataServiceId: "redstone-main-demo",
+          uniqueSignersCount: 1,
+          dataFeeds: dataFeedIdsString,
+          disablePayloadsDryRun: true,
+        },
+        [
+          "https://oracle-gateway-1.b.redstone.finance",
+          "https://d33trozg86ya9x.cloudfront.net",
+        ]
+      );
+      break;
+    default:
+      return {
+        canExec: false,
+        message: `Data service id not found: ${dataServiceId}`,
+      };
+  }
+
+  // Retrieve live prices
+  const { data: livePriceData } =
+    await wrappedOracle.populateTransaction.getLivePricesAndTimestamp(
+      dataFeedIdsBytes32
+    );
+  const txCalldataBytes = arrayify(String(livePriceData));
+  const parsingResult = redstone.RedstonePayload.parse(txCalldataBytes);
```

Comment:
This should be within a `try..catch` in case parsing fails and the `assertValidRedstoneMarker` assertion is triggered.

```suggestion
  const parsedResult = redstone.RedstonePayload.parse(txCalldataBytes);
```

Reply:
Will be addressed in separate [Issue](https://github.com/LiskHQ/lisk-contracts/issues/210)